### ISSUE TITLE
Unable to open more than ±15-20 simultaneous connections to Windows 2012 R2 and higher

### ISSUE THREAD

**Issue by Quinfox** [mRemoteNG/mRemoteNG#864]
When opening 15-20 simultaneous RDP connections to Windows 2012 R2 or higher, new connections are not made and return error 3334 1-2 seconds after filling in credentials and connecting. New connections to Windows 2008 R2 servers remain succesfull. All connections to Windows 2012 R2 and higher servers in different networks, domains etc. are unsuccesfull. New connections to Windows 2012 R2 and higher servers using manual mstsc.exe works fine. New connections to Windows 2012 R2 and higher servers using a new instance of mRemoteNG work fine too, but the same limit applies to the new instance.

|Detail|Value|
|--:|---|
Operating system| Windows 2012 R2 x64
mRemoteNG version| 1.75.7012.16814

**Comment by mmetully** [comment]
This behavior is not mRemoteNG specific. There a partial fix, but ultimately 2012 R2+ rdp connections use enough memory that the virtual memory runs out of space. There are two visual studio tool commands that can mostly fix this.

https://social.technet.microsoft.com/Forums/windowsserver/en-US/679406ff-4639-46da-b211-3f3bde6bb47e/remote-desktop-connection-manager-can-only-open-6-sessions-at-a-time-on-server-2012?forum=winserverTS

**Comment by St1ng3r** [comment]
This also happens with Windows Server 2016. Technet article refers to disabling a security setting, don't really like that. For now i open 15 connections with local mremoteng and start two mremoteng thru Published Apps.

**Comment by mmetully** [comment]
Right, and sensible. The particular fix I was focused on was using visual studio tools to enable large address spaces for the management software. I've seen references to using it on rd manager exe and remoteng exe.

**Comment by sparerd** [comment]
@mmetully We already enable large address aware (LAA) on our exe before publishing it. We have looked into providing an x64 build of mRemoteNG as well, but that would prevent us from providing Citrix support (I have only been able to find a 32-bit version of their client library).

**Comment by Quinfox** [comment]
**@mmetully** I tested several other alternative remote desktop management tools (Remote Desktop Connection Manager, Terminals, EasyConnect), and all of them where able to connect to more than 20 servers simultaneously, so it appears to be a mRemoteNG specific issue. The memory load of mRemoteNG was always stuck near 1 GB when the issue happened, but closing existing connections (and freeing memory) made it impossible to start new ones after the issue happens.
Personally I don't use the Citrix client integration, so a x64 (even a beta) build would be appreciated. Either that, or maybe try to spawn the child processes in separate memory?

**Comment by sparerd** [comment]
This might be mitigated by a v1.77 change that I've been spiking that will allow you to select the version of the RDP client to use during connection. Currently, we only support client version 8. Though I think the most recent is v10. This version bump might contain memory usage improvements or other bug fixes.

`[...] or maybe try to spawn the child processes in separate memory?`
We could look into that as well

**Comment by Mr-Jazz** [comment]
RDP v6.1 seems to be "magic" version that worked best in my POC testing.  I successfully downgraded a 2016 server to RDP v6.1 and have been able to open 90+ concurrent RDP connections without issue.  I've also been able to install both of the old 1.5 and 1.72 mR versions running in tandem on the 2016 POC server and both are working flawlessly so far.

**Comment by Mr-Jazz** [comment]
A few more notes from my POC testing with mR 1.50 and mRNG 1.72 installed on Server 2016:
mstsc-v6.1.7600.16385 – 102 concurrent RDP connections in NG 1.72
mstsc-v6.1.7601.24514 – 94  concurrent RDP connections in NG 1.72

Going back to v6.0.6002.18356 causes the NLA failure (error 2825) in NG v1.72
Going forward  beyond 6.1.7601.24514 causes the "A Website is trying to start a remote connection" nag warning in mR v1.50

Running these two mR versions in tandem with RDP v6.1.7600.16385 i'm able to open 200+ connections with old devices in 1.50 and new devices in 1.72

**Comment by Mr-Jazz** [comment]
Based on my observations, the farther back you go in RDP versions - the higher the concurrent RDP count.

Question:
is it possible to implement NLA in mRNG using RDP v6.0.6001.18000?

**Comment by Mr-Jazz** [comment]
Update:
Feb 2020 updates pushed a new version of mstscax.dll - 10.0.14393.3503
I had to revert back to my v6 dll.

**Comment by simonai1254** [comment]
There seems to be a x64 Build planned (see #915) which might as well help to increase the amount of connections due to the assumed RAM limitations...
